Failure cluster [56b62216...] Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion

### Failure cluster [56b622161ccf52c0d3a6](https://go.k8s.io/triage#56b622161ccf52c0d3a6)

##### Error text:
```
Failed
=== RUN   Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion
    gc_controller_test.go:127: unexpected storage version object
    gc_controller_test.go:128: got: &StorageVersion{ObjectMeta:{k8s.test.resources      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:StorageVersionSpec{},Status:StorageVersionStatus{StorageVersions:[]ServerStorageVersion{ServerStorageVersion{APIServerID:kube-apiserver-1,EncodingVersion:v1,DecodableVersions:[v1],ServedVersions:[],},ServerStorageVersion{APIServerID:kube-apiserver-2,EncodingVersion:v2,DecodableVersions:[v2],ServedVersions:[],},ServerStorageVersion{APIServerID:kube-apiserver-3,EncodingVersion:v2,DecodableVersions:[v2],ServedVersions:[],},},CommonEncodingVersion:*v1,Conditions:[]StorageVersionCondition{},},}
    gc_controller_test.go:129: expected: [{APIServerID:kube-apiserver-2 EncodingVersion:v2 DecodableVersions:[v2] ServedVersions:[]} {APIServerID:kube-apiserver-3 EncodingVersion:v2 DecodableVersions:[v2] ServedVersions:[]}]
    gc_controller_test.go:133: unexpected common encoding version
    gc_controller_test.go:134: got: "v1"
    gc_controller_test.go:135: expected: "v2"
    gc_controller_test.go:139: expected 1 condition, got: 0
--- FAIL: Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion (2.01s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 137 [running]:
testing.tRu
```
#### Recent failures:
[5/11/2025, 5:11:18 AM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit/1921401449628045312)
[5/6/2025, 7:18:04 AM ci-kubernetes-unit-1-33](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-unit-1-33/1919622457946279936)


/kind failing-test
/kind flake
/sig api-machinery

Managed to reproduce it, but only once:
```
kubernetes $ go test -c ./pkg/controller/storageversiongc/ 
kubernetes $ stress ./storageversiongc.test --test.run Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion
5s: 48 runs so far, 0 failures
10s: 96 runs so far, 0 failures
15s: 168 runs so far, 0 failures

/tmp/go-stress-20250516T114058-1571804413
I0516 11:41:13.330882 1419199 shared_informer.go:677] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I0516 11:41:13.331004 1419199 shared_informer.go:677] "Warning: resync period is too small. Changing it to the minimum allowed value" resyncPeriod="100ms" minimumResyncPeriod="1s"
I0516 11:41:13.331041 1419199 gc_controller.go:101] "Starting storage version garbage collector"
--- FAIL: Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion (2.00s)
    gc_controller_test.go:127: unexpected storage version object
    gc_controller_test.go:128: got: &StorageVersion{ObjectMeta:{k8s.test.resources      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:StorageVersionSpec{},Status:StorageVersionStatus{StorageVersions:[]ServerStorageVersion{ServerStorageVersion{APIServerID:kube-apiserver-1,EncodingVersion:v1,DecodableVersions:[v1],ServedVersions:[],},ServerStorageVersion{APIServerID:kube-apiserver-2,EncodingVersion:v2,DecodableVersions:[v2],ServedVersions:[],},ServerStorageVersion{APIServerID:kube-apiserver-3,EncodingVersion:v2,DecodableVersions:[v2],ServedVersions:[],},},CommonEncodingVersion:*v1,Conditions:[]StorageVersionCondition{},},}
    gc_controller_test.go:129: expected: [{APIServerID:kube-apiserver-2 EncodingVersion:v2 DecodableVersions:[v2] ServedVersions:[]} {APIServerID:kube-apiserver-3 EncodingVersion:v2 DecodableVersions:[v2] ServedVersions:[]}]
    gc_controller_test.go:133: unexpected common encoding version
    gc_controller_test.go:134: got: "v1"
    gc_controller_test.go:135: expected: "v2"
    gc_controller_test.go:139: expected 1 condition, got: 0
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 173 [running]:
testing.tRunner.func1.2({0x2d5c8c0, 0xc0006738d8})
        /usr/local/google/home/siarkowicz/.gvm/gos/go1.24.1/src/testing/testing.go:1734 +0x21c
testing.tRunner.func1()
        /usr/local/google/home/si
…
20s: 216 runs so far, 1 failures (0.46%)
25s: 288 runs so far, 1 failures (0.35%)
```

--- a/pkg/controller/storageversiongc/gc_controller_test.go
+++ b/pkg/controller/storageversiongc/gc_controller_test.go
@@ -18,6 +18,7 @@ package storageversiongc
 
 import (
 	"context"
+	"fmt"
 	"reflect"
 	"testing"
 	"time"
@@ -26,9 +27,11 @@ import (
 	coordinationv1 "k8s.io/api/coordination/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
 	"k8s.io/klog/v2/ktesting"
 	"k8s.io/utils/ptr"
 )
@@ -39,8 +42,12 @@ func setupController(ctx context.Context, clientset kubernetes.Interface) {
 	storageVersionInformer := informerFactory.Internal().V1alpha1().StorageVersions()
 
 	controller := NewStorageVersionGC(ctx, clientset, leaseInformer, storageVersionInformer)
+	informerFactory.Start(ctx.Done())
+	// Using this ensure informer caches are fully populated before starting the controller.
+	if !cache.WaitForCacheSync(ctx.Done(), controller.leasesSynced, controller.storageVersionSynced) {
+		panic("timed out waiting for caches to sync")
+	}
 	go controller.Run(context.Background())
-	informerFactory.Start(nil)
 }
 
 func newKubeApiserverLease(name, holderIdentity string) *coordinationv1.Lease {
@@ -93,21 +100,13 @@ func Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion(t *te
 		},
 	}
 
-	clientset := fake.NewSimpleClientset(lease1, lease2, lease3, storageVersion)
+	clientset := fake.NewClientset(lease1, lease2, lease3, storageVersion)
 	_, ctx := ktesting.NewTestContext(t)
 	setupController(ctx, clientset)
 
 	// Delete the lease object and verify that storage version status is updated
 	if err := clientset.CoordinationV1().Leases(metav1.NamespaceSystem).Delete(context.Background(), "kube-apiserver-1", metav1.DeleteOptions{}); err != nil {
 		t.Fatalf("error deleting lease object: %v", err)
-	}
-
-	// add a delay to ensure controller had a chance to reconcile
-	time.Sleep(2 * time.Second)
-
-	storageVersion, err := clientset.InternalV1alpha1().StorageVersions().Get(context.Background(), "k8s.test.resources", metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("error getting StorageVersion: %v", err)
 	}
 
 	expectedServerStorageVersions := []apiserverinternalv1alpha1.ServerStorageVersion{
@@ -122,29 +121,47 @@ func Test_StorageVersionUpdatedWithAllEncodingVersionsEqualOnLeaseDeletion(t *te
 			DecodableVersions: []string{"v2"},
 		},
 	}
+	var lastErr error
 
-	if !reflect.DeepEqual(storageVersion.Status.StorageVersions, expectedServerStorageVersions) {
-		t.Error("unexpected storage version object")
-		t.Logf("got: %+v", storageVersion)
-		t.Logf("expected: %+v", expectedServerStorageVersions)
-	}
+	// Wait up to 5 seconds, checking every 100ms to ensure controller had a chance to reconcile
+	err := wait.PollUntilContextTimeout(ctx, 100*time.Millisecond, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+		storageVersion, err := clientset.InternalV1alpha1().StorageVersions().Get(
+			ctx, "k8s.test.resources", metav1.GetOptions{},
+		)
+		if err != nil {
+			lastErr = fmt.Errorf("failed to get StorageVersion: %w", err)
+			return false, err
+		}
 
-	if *storageVersion.Status.CommonEncodingVersion != "v2" {
-		t.Errorf("unexpected common encoding version")
-		t.Logf("got: %q", *storageVersion.Status.CommonEncodingVersion)
-		t.Logf("expected: %q", "v2")
-	}
+		if !reflect.DeepEqual(storageVersion.Status.StorageVersions, expectedServerStorageVersions) {
+			lastErr = fmt.Errorf("storage versions mismatch: got %+v, expected %+v", storageVersion.Status.StorageVersions, expectedServerStorageVersions)
+			return false, err
+		}
+		if *storageVersion.Status.CommonEncodingVersion != "v2" {
+			t.Errorf("unexpected common encoding version")
+			t.Logf("got: %q", *storageVersion.Status.CommonEncodingVersion)
+			t.Logf("expected: %q", "v2")
+			return false, err
+		}
+		if len(storageVersion.Status.Conditions) != 1 {
+			lastErr = fmt.Errorf("CommonEncodingVersion mismatch: got %v, expected %q", storageVersion.Status.CommonEncodingVersion, "v2")
+			return false, err
+		}
 
-	if len(storageVersion.Status.Conditions) != 1 {
-		t.Errorf("expected 1 condition, got: %d", len(storageVersion.Status.Conditions))
-	}
+		if storageVersion.Status.Conditions[0].Type != apiserverinternalv1alpha1.AllEncodingVersionsEqual {
+			lastErr = fmt.Errorf("expected condition type 'AllEncodingVersionsEqual', got: %q", storageVersion.Status.Conditions[0].Type)
+			return false, nil
+		}
 
-	if storageVersion.Status.Conditions[0].Type != apiserverinternalv1alpha1.AllEncodingVersionsEqual {
-		t.Errorf("expected condition type 'AllEncodingVersionsEqual', got: %q", storageVersion.Status.Conditions[0].Type)
-	}
+		if storageVersion.Status.Conditions[0].Status != apiserverinternalv1alpha1.ConditionTrue {
+			lastErr = fmt.Errorf("expected condition status 'True', got: %q", storageVersion.Status.Conditions[0].Status)
+			return false, nil
+		}
+		return true, nil
+	})
 
-	if storageVersion.Status.Conditions[0].Status != apiserverinternalv1alpha1.ConditionTrue {
-		t.Errorf("expected condition status 'True', got: %q", storageVersion.Status.Conditions[0].Status)
+	if err != nil {
+		t.Fatalf("controller did not reconcile storage version in time: %v\nLast mismatch: %v", err, lastErr)
 	}
 }
 

